src/lpython/semantics/python_ast_to_asr.cpp

```diff
@@ -5962,6 +6079,45 @@ class BodyVisitor : public CommonVisitor<BodyVisitor> {
                 tmp = make_call_helper(al, fn_div, current_scope, args, "_lpython_str_endswith", loc);
             }
             return;
+        } else if (attr_name == "partition") {
+            
+            /*  
+                str.partition(seperator)        ---->
+
+                Split the string at the first occurrence of sep, and return a 3-tuple containing the part 
+                before the separator, the separator itself, and the part after the separator. 
+                If the separator is not found, return a 3-tuple containing the string itself, followed 
+                by two empty strings.
+            */
+
+            if (args.size() != 1) {
+                throw SemanticError("str.partition() takes one arguments",
+                        loc);
+            }
+
+            ASR::expr_t *arg_seperator = args[0].m_value;
+            ASR::ttype_t *arg_seperator_type = ASRUtils::expr_type(arg_seperator);
+            if (!ASRUtils::is_character(*arg_seperator_type)) {
+                throw SemanticError("str.partition() takes one arguments of type: str",
+                    arg_seperator->base.loc);
+            }
+
+            if(s_var.size() == 0) {
+                throw SemanticError("string to undergo partition cannot be empty",loc);
+            }
+
+            if (ASRUtils::expr_value(arg_seperator) != nullptr) {
+
+                string_create_partition(s_var, arg_seperator, loc, arg_seperator_type);
```

Comment:
This should be called evaluation, and the other function should be called creation.

Furthermore, you should store the compile time evaluation in the value attribute of the FunctionCall, in other words, you should always create a function call, but if the compile time evaluation is available, then store it in the "value" member.

Reply:
How can I store the result of compile time evaluation in value attribute of Function Call?

Reply:
If possible, can you give an example of that as well.

Reply:
I think that `ASR::asr_t* make_call_helper` which I am calling from string partition and compile time evaluation function, automatically calls `make_Function_Call_t` which sets the value of Function Call. Isn't it ?

Reply:
I think @certik is suggested to create `FunctionCall` in `string_create_partition`, as I can see in your implementation, you are creating a `TupleConstant` and assigning that to `tmp`. Instead Make a `FunctionCall` and store the `TupleConstant` in the value attribute and assign `FunctionCall` to `tmp`.

From 

```
    [(TupleConstant
        [(StringConstant
            ""
            (Character 1 1 () [])
        )
        (StringConstant
            "1"
            (Character 1 1 () [])
        )
        (StringConstant
            "234"
            (Character 1 1 () [])
        )]
        (Tuple
            [(Character 1 1 () [])
            (Character 1 1 () [])
            (Character 1 1 () [])]
        )
    )]
```
To 
```
[(FunctionCall
    2 _lpython_str_partition@__lpython_overloaded_0___lpython_str_partition
    2 _lpython_str_partition
    [((Var 2 x))
    ((StringConstant
        "1"
        (Character 1 1 () [])
    ))]
    (Tuple
        [(Character 1 -2 () [])
        (Character 1 -2 () [])
        (Character 1 -2 () [])]
    )
    (TupleConstant
        [(StringConstant
            ""
            (Character 1 1 () [])
        )
        (StringConstant
            "1"
            (Character 1 1 () [])
        )
        (StringConstant
            "234"
            (Character 1 1 () [])
        )]
        (Tuple
            [(Character 1 1 () [])
            (Character 1 1 () [])
            (Character 1 1 () [])]
        )
    )
    ()
)]
```

Reply:
Try this and see if you can implement:
- merge both functions
- add the check `ASRUtils::expr_value(arg_seperator) != nullptr`
- Call `eval_partition` inside the `create_partition` function.
- assign FunctionCall to tmp.

Reply:
I am trying to write `evaluate_partition` like this - 

```
void evaluate_partition(std::string &s_var, ASR::expr_t* arg_seperator, 
        const Location &loc, ASR::ttype_t *arg_seperator_type) {

        ASR::StringConstant_t* seperator_constant = ASR::down_cast<ASR::StringConstant_t>(arg_seperator);
        std::string seperator = seperator_constant->m_s;
        if(seperator.size() == 0) {
            throw SemanticError("empty separator", arg_seperator->base.loc);
        }

        int seperator_pos = KMP_string_match(s_var, seperator);
        Vec<ASR::expr_t *> res_tuple;
        Vec<ASR::ttype_t *> res_tuple_type;
        res_tuple.reserve(al, 3); 
        res_tuple_type.reserve(al, 3);
        std :: string first_res, second_res, third_res;
        if(seperator_pos == -1) {
            /* seperator does not exist */
            first_res = s_var;
            second_res = "";
            third_res = "";
        } else {
            first_res = s_var.substr(0, seperator_pos);
            second_res = seperator;
            third_res = s_var.substr(seperator_pos + seperator.size());
        }

        res_tuple.push_back(al, ASRUtils::EXPR(ASR::make_StringConstant_t(al, loc, s2c(al, first_res), arg_seperator_type)));
        res_tuple.push_back(al, ASRUtils::EXPR(ASR::make_StringConstant_t(al, loc, s2c(al, second_res), arg_seperator_type)));
        res_tuple.push_back(al, ASRUtils::EXPR(ASR::make_StringConstant_t(al, loc, s2c(al, third_res), arg_seperator_type)));
        res_tuple_type.push_back(al, arg_seperator_type);
        res_tuple_type.push_back(al,arg_seperator_type);
        res_tuple_type.push_back(al,arg_seperator_type);

        ASR::ttype_t *tuple_type = ASRUtils::TYPE(ASR::make_Tuple_t(al, loc, res_tuple_type.p, res_tuple_type.n));
        ASR::symbol_t *fn_div = resolve_intrinsic_function(loc, "_lpython_str_partition");
        ASR::expr_t* value = ASRUtils::EXPR(ASR::make_TupleConstant_t(al, loc, res_tuple.p, res_tuple.size(), tuple_type));

        Vec<ASR::call_arg_t> args;
        args.reserve(al, 1);
        ASR::call_arg_t str_arg;
        str_arg.loc = loc;
        ASR::ttype_t *str_type = ASRUtils::TYPE(ASR::make_Character_t(al, loc, 1, s_var.size(), nullptr, nullptr, 0));
        str_arg.m_value = ASRUtils::EXPR(ASR::make_StringConstant_t(al, loc, s2c(al, s_var), str_type));
        ASR::call_arg_t sub_arg;
        sub_arg.loc = loc;
        sub_arg.m_value = arg_seperator;
        args.push_back(al, str_arg);
        args.push_back(al, sub_arg);
        
        tmp = ASR::make_FunctionCall_t(al, loc, fn_div, nullptr, args.p, args.n, tuple_type, value, nullptr);
        return;
    }
```


But it throws this error - 
```
ASR verify pass error: ASR verify: FunctionCall::m_name must be a Function or Variable with FunctionType
Internal Compiler Error: Unhandled exception
Traceback (most recent call last):
  Binary file "/mnt/d/B.Tech/workspace/wsl-workspace/lpython/src/bin/lpython", in _start()
  Binary file "/lib/x86_64-linux-gnu/libc.so.6", in __libc_start_main()
  Binary file "/mnt/d/B.Tech/workspace/wsl-workspace/lpython/src/bin/lpython", in main()
  File "lpython.cpp", line 0, in (anonymous namespace)::compile_python_to_object_file(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, LCompilers::PassManager&, LCompilers::CompilerOptions&, bool, bool)
LCompilersException: Verify failed
```

Reply:
You need to enable stacktraces, and we might need to port from LFortran the functionality to show in source code where the verify() failed.